web/src/app/components/timeline/timeline.component.ts

```diff
@@ -0,0 +1,430 @@
+    // respective start and end time.
+    this.svg = (d3.select(element).append('svg') as d3SVG)
+      .attr('width', this.width + margin.left + margin.right)
+      .attr('height', this.height + margin.top + margin.bottom)
+      .attr('viewBox', `0 0 ${this.width} ${this.height}`)
+      .attr('preserveAspectRatio', 'xMinYMin') // Keep aspect ratio on resize
+      .attr('transform', `translate(${margin.left} ${margin.top})`);
+
+    this.svg.call(zoom);
+
+    this.svg.append('g')
+      .attr('class', 'x-axis')
+      .attr('transform', `translate(0 ${this.height})`)
+      .call(xAxis)
+      .selectAll('line')
+      .style('stroke', TimelineComponent.COLOR_LIGHT_GRAY);
+
+    this.svg.append('defs') // Determine what can be seen of the component
+      .append('clipPath')
+      .attr('id', 'chart-content')
+      .append('rect')
+      .attr('x', 0)
+      .attr('y', 0)
+      .attr('height', this.height)
+      .attr('width', this.width);
+
+    this.svg.append('rect')
+      .attr('class', 'chart-bounds o-025')
+      .attr('x', 0)
+      .attr('y', 0)
+      .attr('height', this.height)
+      .attr('width', this.width);
+
+    // Insert timeline interval bars with their y-position determined by their
+    // row index. In this case, `selectAll()` returns an empty selection since
+    // the class does not yet exist in the page. `data(this.data)` and `enter()`
+    // subsequently attaches data to the selection, with the first item in
+    // data corresponding to the first slot in the selection, and so on.
+    const groupHeight = this.height / this.numRows;
+    const groupIntervalItems = this.svg.selectAll('.group-interval-item')
+      .data(this.data)
+      .enter()
+      .append('g')
+      .attr('clip-path', 'url(#chart-content)')
+      .attr('transform', (d: Item) => `translate(0, ${groupHeight * d.row})`)
+      .attr('style', (d: Item) =>
+        `fill: ${TimelineComponent.STATE_TO_COLOR[d.state]}`);
+
+    const intervalBarHeight = 0.8 * groupHeight; // Space between each interval
+    const intervalBarMargin = (groupHeight - intervalBarHeight) / 2;
+    groupIntervalItems
+      .append('rect')
+      .attr('class', 'interval pointer')
+      .attr('width', (d: Item) => this.x(d.endTime) - this.x(d.startTime))
+      .attr('height', intervalBarHeight)
+      .attr('rx', 2)
+      .attr('ry', 2)
+      .attr('y', intervalBarMargin)
+      .attr('x', (d: Item) => this.x(d.startTime));
+
+    // Create the tooltip and set its opacity to 0 when not hovering over a
+    // set of data, such that it only appears when the cursor is directly on top
+    // of an interval.
+    const tooltipDiv = document.createElement('div');
+    const tooltip = d3.select(tooltipDiv).call(this.styleTooltip);
+    element.appendChild(tooltipDiv);
+
+    groupIntervalItems
+      .on('mouseover', (d: Item) => {
+        d3.select(d3.event.currentTarget)
+          .select('rect')
+          .attr('fill-opacity', 0.50);
+
```

Comment:
I'd remove this empty line here since it seems a bit easier to read Line#401-Line#416.